### PR TITLE
v1beta2 APIs

### PR DESCRIPTION
**Pull request by sedefsavas**
**What this PR does / why we need it**:
This PR adds new v1beta2 API version and it is the exact copy of v1beta1.
After this PR is merged, v1beta2 changes will come as follow up PRs.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Checklist**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

### CODE REVIEW

**Comment by Ankitasw** [comment]
> This looks okay to me compared to previous ways of promoting a new version. But since this is a significant change, I would suggest not just taking my word for it. :D

Ahh and it just happened, will fix E2E tests.
@Skarlso All the very best in reviewing this PR though 😛

**Comment by Skarlso** [comment]
>  @Skarlso All the very best in reviewing this PR though 😛

Bwuhahahaha. :D Yeah, it's quite massive. :D

**Comment by Ankitasw** [comment]
Running E2E tests locally, and would commit the fix together, mostly failing because of some of the kubeadm APIs are still v1beta1.

**Comment by sedefsavas** [comment]
/test pull-cluster-api-provider-aws-e2e-eks



<br class="Apple-interchange-newline">

**Comment by sedefsavas** [comment]
/test pull-cluster-api-provider-aws-e2e-eks
Some boskos error.

**Comment by Ankitasw** [comment]
@Skarlso @richardcase could you please take some time to review this PR?

**Comment by Skarlso** [comment]
I will take a look today. :)

File: Makefile [hunk 1]
```diff
@@ -214,66 +213,46 @@ generate-go-apis: ## Alias for .build/generate-go-apis
...
 		paths=./cmd/... \
 		object:headerFile=./hack/boilerplate/boilerplate.generatego.txt
 
 	$(MAKE) defaulters
 
 	$(CONVERSION_GEN) \
-		--input-dirs=./api/v1alpha3 \
-		--input-dirs=./api/v1alpha4 \
+		--input-dirs=./api/v1beta1 \
 		--input-dirs=./cmd/clusterawsadm/api/bootstrap/v1alpha1 \
```
Comment (Skarlso) [inline Makefile:230]: I think this folder was also removed, not? Or is that bootstrap/eks... 🤔

Comment (Ankitasw) [reply Makefile:230]: Are you talking about `./cmd/clusterawsadm/api/bootstrap/v1alpha1 `? It is still there.

Comment (Skarlso) [reply Makefile:230]: Yeah, what I was looking at was bootstrap/eks indeed. That's the one that is gone. This is fine. :)

**Comment by Skarlso** [comment]
/lgtm

I would prefer a second set of eyes. :) As far as I can tell the tests are passing and the code seems to be okay all alpha versions have been deprecated and the mention of them. So... I'm good. :D

**Comment by Ankitasw** [comment]
> I would prefer a second set of eyes. :) As far as I can tell the tests are passing and the code seems to be okay all alpha versions have been deprecated and the mention of them. So... I'm good. :D

We will wait for @richardcase to approve 🙂

**Comment by richardcase** [comment]
So e2e-blocking and e2e-eks are passing, thats good. Looking now.....

**Review by richardcase — COMMENTED** [review]
I wonder if we've allowed enough time for people to upgrade to v1beta1? But i guess we won't do a v1beta2 release for a while.

File: config/crd/kustomization.yaml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 commonLabels:
   cluster.x-k8s.io/v1alpha3: v1alpha3
   cluster.x-k8s.io/v1alpha4: v1alpha4
-  cluster.x-k8s.io/v1beta1: v1beta1
+  cluster.x-k8s.io/v1beta1: v1beta1_v1beta2
```
[the remark this answers is not in the document]
Comment (sedefsavas) [reply config/crd/kustomization.yaml:4]: Yes, this is convention, e.g., https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/5723804e2284244dff30572edd40f023442779ad/config/crd/kustomization.yaml#L3

File: api/v1beta2/bastion.go [hunk 3]
```diff
@@ -43,3 +48,16 @@ func (b *Bastion) Validate() []*field.Error {
 	}
 	return errs
 }
+
+func validateSSHKeyName(sshKeyName *string) field.ErrorList {
```
[the remark this answers is not in the document]
Comment (sedefsavas) [reply api/v1beta2/bastion.go:52]: I agree, added this as todo to v1beta2 API: https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/2355#issuecomment-1248386090

**Comment by Ankitasw** [comment]
@richardcase @Skarlso This is ready for final review.

**Comment by luthermonson** [comment]
🚀

File: Makefile [hunk 4]
```diff
@@ -214,66 +213,46 @@ generate-go-apis: ## Alias for .build/generate-go-apis
 		paths=./bootstrap/eks/api/... \
 		paths=./controlplane/eks/api/... \
 		paths=./iam/api/... \
-		crd:crdVersions=v1 \
 		output:crd:dir=config/crd/bases \
-		output:webhook:dir=$(WEBHOOK_ROOT) \
 		object:headerFile=./hack/boilerplate/boilerplate.generatego.txt \
+		crd:crdVersions=v1 \
+		rbac:roleName=manager-role \
 		webhook
```
Comment (richardcase) [inline Makefile:221]: We seem to have lost the controller-gen on the `controllers` folders...i think this needs to be added back in.

File: api/v1beta2/bastion.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (richardcase) [reply api/v1beta2/bastion.go:52]: Great, thanks.

File: config/crd/kustomization.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (richardcase) [reply config/crd/kustomization.yaml:4]: Nice :+1:

File: Makefile (hunk 4 above)
Comment (richardcase) [reply Makefile:221]: Apart from ^^^ that this looks good.

Comment (Ankitasw) [reply Makefile:221]: Do we need controllers folder in controller-gen? Is there any generated code in that folder? I actually didn't find any.

Comment (richardcase) [reply Makefile:221]: All of the RBAC generation comes from the controllers

Comment (richardcase) [reply Makefile:221]: So its needed :)

Comment (Ankitasw) [reply Makefile:221]: ok :) Will add it

Comment (Ankitasw) [reply Makefile:222]: Added, PTAL

**Comment by richardcase** [comment]
Based on review, discussions and e2e passing:

/approve

**Pull request merged**